h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,377 @@
+        if (_iTrees != null) {
+            out.iTrees = _iTrees;
+        }
+    }
+
+    private class ExtendedIsolationForestDriver extends Driver {
+
+        @Override
+        protected ExtendedIsolationForestModel makeModel(Key<ExtendedIsolationForestModel> modelKey,
+                                                         ExtendedIsolationForestModel.ExtendedIsolationForestParameters parms) {
+            return new ExtendedIsolationForestModel(modelKey, parms, new ExtendedIsolationForestModel.ExtendedIsolationForestOutput(ExtendedIsolationForest.this));
+        }
+
+        @Override
+        protected boolean doOOBScoring() {
+            return true;
+        }
+
+        @Override
+        protected void scoreAndBuildTrees(boolean oob) {
+            IsolationTreeForkJoinTask [] iTreeTasks = new IsolationTreeForkJoinTask[_parms._ntrees];
+            for (int t = 0; t < _parms._ntrees; t++) {
+                iTreeTasks[t] = new IsolationTreeForkJoinTask();
+                H2O.submitTask(iTreeTasks[t]);
+            }
+            for (int t = 0; t < _parms._ntrees; t++) {
+                _iTrees[t] = iTreeTasks[t].getResult();
+            }
+            addCustomInfo(_model._output);
+        }
+
+        @Override
+        protected boolean buildNextKTrees() {
+            // this method is not used in the current implementation
+            // tree building is implemented in scoreAndBuildTrees method
+            return false;
+        }
+
+        @Override
+        protected void initializeModelSpecifics() {
+            _iTrees = new IsolationTree[_parms._ntrees];
+        }
+
+    }
+
+    /**
+     * IsolationTree class implements Algorithm 2 (iTree) and Algorithm 3 (pathLength) 
+     * from the Extended Isolation Forest paper.
+     * 
+     * The binary tree recursive implementation is replaced by array implementation.
+     * It is the most simple array implementation of binary tree.
+     */
+    public static class IsolationTree extends Iced<IsolationTree> {
+        private Node[] _nodes;
+
+        private Key<Frame> _frameKey;
+        private int _heightLimit;
+        private long _seed;
+        private int _extensionLevel;
+
+        public IsolationTree(Key<Frame> frame, int _heightLimit, long _seed, int _extensionLevel) {
+            this._frameKey = frame;
+            this._heightLimit = _heightLimit;
+            this._seed = _seed;
+            this._extensionLevel = _extensionLevel;
+
+            int maxNumNodesInTree = (int) Math.pow(2, _heightLimit) - 1;
+            this._nodes = new Node[maxNumNodesInTree]; 
+        }
+
+        /**
+         * Implementation of Algorithm 2 (iTree) from paper.
+         * Algorithm not need exact row vectors in scoring (evaluation) stage but just number of rows in node.
+         * Therefore nodeFrames are removed from DKV after filtering.
+         */
+        public void buildTree() {
+            Frame frame = DKV.get(_frameKey).get();
```

Comment:
seems unnecessary, can be moved to Node constructor and just do getFrame().numRows()

Reply:
Will be fixed in https://github.com/h2oai/h2o-3/pull/4907